libbeat/otelbeat/oteltest/oteltest.go

```diff
@@ -0,0 +1,115 @@
+// specific language governing permissions and limitations
+// under the License.
+
+// Package oteltest provides test utilities for OpenTelemetry and Beats components.
+package oteltest
+
+import (
+	"bytes"
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/elastic/elastic-agent-libs/mapstr"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/receiver"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type CheckMultipleReceiversParams struct {
+	T *testing.T
+	// Factory that allows to create a receiver.
+	Factory receiver.Factory
+	// Config of the receiver to use.
+	Config component.Config
+	// AssertFunc is a function that asserts the test conditions.
+	AssertFunc func(t *testing.T, logs map[string][]mapstr.M)
+}
+
+// CheckMultipleReceivers checks that multiple receivers can be created and started
+// on the same process without errors.
+func CheckMultipleReceivers(params CheckMultipleReceiversParams) {
+	t := params.T
+	var logsMu sync.Mutex
+	logs := make(map[string][]mapstr.M)
+
+	ctx := context.Background()
+	createReceiver := func(t *testing.T, name string) receiver.Logs {
+		t.Helper()
+
+		var zapLogs bytes.Buffer
+		core := zapcore.NewCore(
+			zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()),
+			zapcore.AddSync(&zapLogs),
+			zapcore.DebugLevel)
+
+		receiverSettings := receiver.Settings{}
+		receiverSettings.Logger = zap.New(core).Named(name)
+
+		logConsumer, err := consumer.NewLogs(func(ctx context.Context, ld plog.Logs) error {
+			for i := 0; i < ld.ResourceLogs().Len(); i++ {
+				rl := ld.ResourceLogs().At(i)
+				for j := 0; j < rl.ScopeLogs().Len(); j++ {
+					sl := rl.ScopeLogs().At(j)
+					for k := 0; k < sl.LogRecords().Len(); k++ {
+						log := sl.LogRecords().At(k)
+						logsMu.Lock()
+						logs[name] = append(logs[name], log.Body().Map().AsRaw())
+						logsMu.Unlock()
+						t.Logf("ingested log for %q: %v", name, log.Body().Map().AsRaw())
+					}
+				}
+			}
+			return nil
+		})
+		assert.NoErrorf(t, err, "Error creating log consumer for %q", name)
+
+		t.Cleanup(func() {
+			if t.Failed() {
+				t.Logf("Logs for %q: %s\n", name, zapLogs.String())
+			}
+		})
+
+		r, err := params.Factory.CreateLogs(ctx, receiverSettings, params.Config, logConsumer)
+		assert.NoErrorf(t, err, "Error creating receiver %q", name)
+		return r
+	}
+
+	r1 := createReceiver(t, "r1")
+	r2 := createReceiver(t, "r2")
+
+	err := r1.Start(ctx, nil)
+	require.NoError(t, err, "Error starting receiver 1")
+	defer func() {
+		require.NoError(t, r1.Shutdown(ctx), "Error shutting down receiver 1")
+	}()
+
+	err = r2.Start(ctx, nil)
+	require.NoError(t, err, "Error starting receiver 2")
+	defer func() {
+		require.NoError(t, r2.Shutdown(ctx), "Error shutting down receiver 2")
+	}()
+
+	logsMu.Lock()
+	defer logsMu.Unlock()
```

Comment:
You're locking `logsMu` until `AssertFunc` returns, which means you're locking `logsMu` until the test ends.

How can you be sure both receivers have already written to `logs`/do not need the lock any more?

I could see a case where only one of the receivers has started and `AssertFunc` is called, because the lock is being held outside of `AssertFunc`, then the receivers can never get the lock, effectively dead-locking the test.

Reply:
I noticed this, it is kinda like a cat and mouse problem. The simplest solution I came up with is to call the assert func inside a `require.Eventually` so we only lock for the duration of the callback which allows the consumer to make progress while not blocking the assertions and vice-versa. Let me knnow if you have a better idea in mind.